### PR TITLE
Lwm2m swmgmt

### PR DESCRIPTION
**Pull request by DunderRoffe**

### CODE REVIEW

File: include/net/lwm2m.h [hunk 1]
```diff
@@ -318,6 +319,76 @@ struct coap_block_context *lwm2m_firmware_get_block_context();
 #endif
 #endif
 
+
+#if defined(CONFIG_LWM2M_SWMGMT_OBJ_SUPPORT)
```
Comment (endian-benjamin) [inline include/net/lwm2m.h:323]: Seems wrong to have these prototypes inside the lwm2m header, doesn't seem as if any other objects do it like this?

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c [hunk 2]
```diff
@@ -27,13 +27,20 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
 
-static struct k_work firmware_work;
-static char firmware_uri[URI_LEN];
-static struct lwm2m_ctx firmware_ctx = {
-	.sock_fd = -1
+struct firmware_pull_context {
+	char uri[URI_LEN];
+	int retry;
+	struct k_work firmware_work;
+	struct lwm2m_ctx firmware_ctx;
```
Comment (endian-benjamin) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: prob rename this field to lwm2m_ctx or similar

File: subsys/net/lib/lwm2m/lwm2m_pull_context.c [hunk 3]
```diff
@@ -0,0 +1,467 @@
+#define LOG_MODULE_NAME net_lwm2m_pull_context
+//#define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
+#define LOG_LEVEL LOG_LEVEL_DBG
```
Comment (endian-benjamin) [inline subsys/net/lib/lwm2m/lwm2m_pull_context.c:3]: ^----

File: subsys/net/lib/lwm2m/Kconfig [hunk 4]
```diff
@@ -215,6 +215,17 @@ config LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 	  block transfer and "FIRMWARE PACKAGE URI" resource.  This option
 	  adds another UDP context and packet handling.
 
+config LWM2M_SWMGMT_OBJ_SUPPORT
+	bool "Software management object support"
+	default y
+	help
+	  Include support for LWM2M Software Management Object (ID 9)
+
+config LWM2M_SWMGMT_MAX_INSTANCE_COUNT
```
Comment (endian-benjamin) [inline subsys/net/lib/lwm2m/Kconfig:224]: condition on SWMGTM_OBJ_SUPPORT, choose a different default prob

File: subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+
+	if (!cb) {
+		return -EINVAL;
+	}
+
+	instance = find_index(obj_inst_id);
+	if (!instance) {
+		return -EINVAL;
```
Comment (endian-benjamin) [inline subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c:268]: ENOENT?

File: subsys/net/lib/lwm2m/Kconfig (hunk 4 above)
Comment (DunderRoffe) [reply subsys/net/lib/lwm2m/Kconfig:224]: I thought this was solved by "depends on LWM2M_SWMGMT_OBJ_SUPPORT" below?

File: include/net/lwm2m.h (hunk 1 above)
Comment (DunderRoffe) [reply include/net/lwm2m.h:323]: FOTA does this just above... Copied it from there

**Comment by JoshuaJeyaraj** [comment]
https://docs.google.com/spreadsheets/d/1sjT5Kc5ppcTw8WSy-7MW0bAh8wYRG2vvqUGvnNTYNhs/edit#gid=0

Test case list

**Comment by DunderRoffe** [comment]
Closing and continuing in https://github.com/voiapp/zephyr/pull/8

**Pull request closed without merging**